keps/sig-auth/3221-structured-authorization-configuration/README.md

```diff
@@ -0,0 +1,615 @@
+
+
+## Design Details
+
+We would like to introduce a structured file format which allows authorization
+to be configured using a flag (`--authorization-config-file`) which accepts a
+path to a file on the disk. This feature can be enabled or disabled by the
+explicit feature flag `AuthorizationConfigFromFile`.
+The proposed structure is illustrated below:
+
+```yaml
+apiVersion: apiserver.config.k8s.io/v1alpha1
+kind: AuthorizationConfiguration
+authorizers:
+  - type: Webhook
+    webhook:
+    unauthorizedTTL: 30s
+    timeout: 3s
+    subjectAccessReviewVersion: v1
+    onError: Deny
+    connectionInfo:
+      type: KubeConfig
+      kubeConfigFile: /kube-system-authz-webhook.yaml
+    matchConditions:
+    - expression: |
+      request.resourceAttributes.namespace == 'kube-system'
+    - expression: |
+      !('system:serviceaccounts:kube-system' in request.user.groups)
+  - type: Node
+  - type: RBAC
+  - type: Webhook
+    webhook:
+    authorizedTTL: 5m
+    unauthorizedTTL: 30s
+    timeout: 3s
+    subjectAccessReviewVersion: v1
+    onError: NoOpinion
+    connectionInfo:
+      type: InClusterConfig
+```
+
+Validation will allow multiple authorizers of type "Webhook" to be added to the
+config, but one authorizer each for other types. The ordering of this chain will
+be decided by the order specified in the file.
+
+The keys `kubeConfigFile`, `authorizedTTL`, `unauthorizedTTL` and
+`subjectAccessReviewVersion` accept values corresponding to flags
+`--authorization-webhook-config-file`, `--authorization-webhook-cache-authorized-ttl`,
+`--authorization-webhook-cache-unauthorized-ttl` and `--authorization-webhook-version`
+respectively.
+
+This diagram shows how the configuration flags map to configuration file keys
+for a webhook:
+![](https://i.imgur.com/PJSmgOq.jpg)
+
+The `onError` will be a required field which allow users to specify whether
+or not request is denied if the webhook errors out or is unreachable. This allows
+`SubjectAccessReviews` to be more definitive.
+
+Today, the `SubjectAccessReview` version defaults to `v1beta1` if the corresponding
+flag is not supplied. While configuring authorization modes using the file config,
+the version supported by a webhook has to be mentioned using a required field
+`subjectAccessReviewVersion`.
+
+The new structure introduces a configurable timeout that allows administrators to
+define a timeout for webhooks in case they are unreachable.
+
+> Note: ABAC will not be supported by the new file format as we discourage the usage of this mode.
+
+The code path for enabling the above will only be triggered if the feature flag will
+be enabled until the time the feature flag is removed and configuring authorizer
+through a file becomes GA.
+
+> TODO: More details on the implementation will be charted once the goals, non-goal and proposal is approved.
+
+### Monitoring
+
+We will add the following 4 metrics:
+
+1. `apiserver_authorization_step_invocations_total`
```

Comment:
How about if we follow `pod_security_evaluations_total` 
> Number of policy evaluations that occurred, not counting ignored or exempt requests.

We can also exclude `NoOpinion` in the count?

Reply:
Per discussion with @enj, he is ok with 
> We can also exclude NoOpinion in the count

and prefers `apiserver_authorization_decisions_total`

Reply:
I like both the suggested names. Going with `apiserver_authorization_decisions_total` since that seems to be the mutual preference.

Also, dropped `NoOpinion` in the count.